File: MiniGame-Java/src/main/java/rs/lab/mges/game/RabbitGame.java
package rs.lab.mges.game;

import io.github.libsdl4j.api.render.SDL_Renderer;
import rs.lab.mges.engine.AssetsLoader;
import rs.lab.mges.engine.Game;
import rs.lab.mges.engine.Sound;

public class RabbitGame extends Game {

    public enum GameScene {
        TITLE,
        GAMEPLAY
    };

    public GameAssets assets;

    private Scene currentScene;
    private TitleScene titleScene;
    private GameplayScene gameplayScene;

    public RabbitGame() {
        super(640, 480, "Rabbit unleashed", 0);

        /* setup assets folder relative to the current folder */
        AssetsLoader.setAssetsFolder("assets");
    }

    @Override
    protected void init(SDL_Renderer renderer) {

        assets = new GameAssets(renderer);
        titleScene = new TitleScene(this);
        gameplayScene = new GameplayScene(this);
        currentScene = titleScene;

        Sound.playMusicFromMemory(assets.music);
    }

    @Override
    protected void free() {
        assets.freeAll();
    }

    public void changeScene(GameScene scene) {
        var newScene = switch (scene) {
            case TITLE ->
                titleScene;
            case GAMEPLAY ->
                gameplayScene;
            default ->
                throw new RuntimeException("Unexpected scene");
        };

        if (currentScene.equals(newScene)) {
            return;
        }

        currentScene.leave();
        currentScene = newScene;
        currentScene.enter();
    }

    @Override
    protected void draw(SDL_Renderer renderer, float interpolationAlpha) {
        currentScene.draw(renderer, interpolationAlpha);
    }

    @Override
    protected void update(float dt) {
        currentScene.update(dt);
    }
}

File: MiniGame-Python/game/RabbitGame.py
from engine import Game, Sound, Text
from .TitleScene import TitleScene
from .GameplayScene import GameplayScene
from .Assets import Assets
from .Scene import GameScene
from game import Scene

__all__ = ['RabbitGame', 'GameScene']


class RabbitGame(Game):
    def __init__(self):
        super().__init__(640, 480, title="Rabbit unleashed")
        self._currentScene: Scene | None = None

    def init(self, renderer):
        # Init text rendering
        Text.init()
        
        self._sound = Sound.initAudio()
        self.assets = Assets(renderer, self._sound)
        self._currentScene = TitleScene(self)
        self._sound.playMusicFromMemory(self.assets.music)  # type: ignore

    def free(self):
        del self.assets

    def changeScene(self, scene):
        newScene = None
        if scene == GameScene.TITLE:
            newScene = TitleScene(self)
        elif scene == GameScene.GAMEPLAY:
            newScene = GameplayScene(self)

        if self._currentScene == newScene:
            return

        self._currentScene.leave()  # type: ignore
        self._currentScene = newScene
        self._currentScene.enter()  # type: ignore

    def update(self, dt):
        if self._currentScene != None:
            self._currentScene.update(dt)

    def draw(self, renderer, delta):
        if self._currentScene != None:
            self._currentScene.draw(renderer, delta)
